Panic using `--libp2p 0.0.0.0:port` option

## Panic using `--libp2p` option

### Description

Running `polygon-sdk server --libp2p 0.0.0.0:10001` fails without passing a `--nat`

### Your environment

* OS and version Ubuntu 20
* version of the Polygon SDK 96be90487c144f7b1589ac65aad3f741c81644e3
* branch that causes this issue `develop`

### Steps to reproduce

* Tell us how to reproduce this issue <br /> Run `polygon-sdk server --libp2p 0.0.0.0:10001` with a genesis file 
* Where the issue is, if you know <br /> `AddrInfoToString` seems to not like when there are two addresses in the libp2p address info. It's giving me two addresses, 

```go
// AddrInfoToString converts an AddrInfo into a string representation that can be dialed from another node
func AddrInfoToString(addr *peer.AddrInfo) string {
	fmt.Print(addr)
	if len(addr.Addrs) != 1 {
		panic("Not supported")
	}
	return addr.Addrs[0].String() + "/p2p/" + addr.ID.String()
}

```
Gives
```
{16Uiu2HAm7dBcv1avSV8TVFZQV4kUNYi5N1jHu7C4xH1gV58McZx2: [/ip4/192.168.1.248/tcp/10002 /ip4/127.0.0.1/tcp/10002]}

```

### Logs

```
panic: Not supported

goroutine 1 [running]:
github.com/0xPolygon/polygon-sdk/network.AddrInfoToString(0xc0005a04e0, 0x0, 0x0)
	/home/julian/sx-network/polygon-sdk/network/server.go:595 +0x1f7
github.com/0xPolygon/polygon-sdk/network.NewServer(0x1582748, 0xc0002f4cb0, 0xc000100eb0, 0x0, 0x0, 0x0)
	/home/julian/sx-network/polygon-sdk/network/server.go:138 +0xe71
github.com/0xPolygon/polygon-sdk/server.NewServer(0x1582748, 0xc0002f4b60, 0xc0002f8f80, 0x0, 0x0, 0x0)
	/home/julian/sx-network/polygon-sdk/server/server.go:92 +0x5ef
github.com/0xPolygon/polygon-sdk/command/server.(*ServerCommand).Run(0xc00000ed08, 0xc000032260, 0x4, 0x4, 0x0)
	/home/julian/sx-network/polygon-sdk/command/server/server_command.go:187 +0x3de
github.com/mitchellh/cli.(*CLI).Run(0xc00013f7c0, 0x0, 0x0, 0x0)
	/home/julian/sx-network/polygon-sdk/vendor/github.com/mitchellh/cli/cli.go:255 +0xb7d
main.Run(0xc000032250, 0x5, 0x5, 0x0)
	/home/julian/sx-network/polygon-sdk/main.go:25 +0x151
main.main()
	/home/julian/sx-network/polygon-sdk/main.go:12 +0x7c
```

### Proposed solution

If you pass a `--nat`, this error disappears. I feel like this should not be a panic.

--- a/network/server.go
+++ b/network/server.go
@@ -603,21 +603,26 @@ var (
 
 // AddrInfoToString converts an AddrInfo into a string representation that can be dialed from another node
 func AddrInfoToString(addr *peer.AddrInfo) string {
-	dialAddress := ""
+	// Safety check
+	if len(addr.Addrs) == 0 {
+		panic("No dial addresses found")
+	}
 
-	// Find an address that's not a loopback address
-	for _, address := range addr.Addrs {
-		if !loopbackRegex.MatchString(address.String()) {
-			// Not a loopback address, dial address found
-			dialAddress = address.String()
-			break
+	dialAddress := addr.Addrs[0].String()
+
+	// Try to see if a non loopback address is present in the list
+	if len(addr.Addrs) > 1 && loopbackRegex.MatchString(dialAddress) {
+		// Find an address that's not a loopback address
+		for _, address := range addr.Addrs {
+			if !loopbackRegex.MatchString(address.String()) {
+				// Not a loopback address, dial address found
+				dialAddress = address.String()
+				break
+			}
 		}
 	}
 
-	if dialAddress == "" {
-		panic("Unable to find appropriate dial address")
-	}
-
+	// Format output and return
 	return dialAddress + "/p2p/" + addr.ID.String()
 }
 

--- a/network/server.go
+++ b/network/server.go
@@ -589,18 +589,20 @@ func StringToAddrInfo(addr string) (*peer.AddrInfo, error) {
 	return addr1, nil
 }
 
-// AddrInfoToString converts an AddrInfo into a string representation that can be dialed from another node
-func AddrInfoToString(addr *peer.AddrInfo) string {
+var (
 	// Regex used for matching loopback addresses (IPv4 and IPv6)
 	// This regex will match:
 	// /ip4/localhost/tcp/<port>
 	// /ip4/127.0.0.1/tcp/<port>
 	// /ip4/<any other loopback>/tcp/<port>
 	// /ip6/<any loopback>/tcp/<port>
-	loopbackRegex := regexp.MustCompile(
+	loopbackRegex = regexp.MustCompile(
 		`/^\\/ip4\\/127(?:\\.[0-9]+){0,2}\\.[0-9]+\\/tcp\\/\\d+$|^\\/ip4\\/localhost\\/tcp\\/\\d+$|^\\/ip6\\/(?:0*\\:)*?:?0*1\\/tcp\\/\\d+$/gm`,
 	)
+)
 
+// AddrInfoToString converts an AddrInfo into a string representation that can be dialed from another node
+func AddrInfoToString(addr *peer.AddrInfo) string {
 	dialAddress := ""
 
 	// Find an address that's not a loopback address

--- a/network/server.go
+++ b/network/server.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"fmt"
 	"net"
+	"regexp"
 	"sync"
 	"time"
 
@@ -14,7 +15,7 @@ import (
 	"github.com/libp2p/go-libp2p-core/host"
 	"github.com/libp2p/go-libp2p-core/network"
 	"github.com/libp2p/go-libp2p-core/peer"
-	peerstore "github.com/libp2p/go-libp2p-core/peerstore"
+	"github.com/libp2p/go-libp2p-core/peerstore"
 	"github.com/libp2p/go-libp2p-core/protocol"
 	noise "github.com/libp2p/go-libp2p-noise"
 	pubsub "github.com/libp2p/go-libp2p-pubsub"
@@ -590,10 +591,31 @@ func StringToAddrInfo(addr string) (*peer.AddrInfo, error) {
 
 // AddrInfoToString converts an AddrInfo into a string representation that can be dialed from another node
 func AddrInfoToString(addr *peer.AddrInfo) string {
-	if len(addr.Addrs) != 1 {
-		panic("Not supported")
+	// Regex used for matching loopback addresses (IPv4 and IPv6)
+	// This regex will match:
+	// /ip4/localhost/tcp/<port>
+	// /ip4/127.0.0.1/tcp/<port>
+	// /ip4/<any other loopback>/tcp/<port>
+	// /ip6/<any loopback>/tcp/<port>
+	loopbackRegex := regexp.MustCompile(
+		"/^\\/ip4\\/127(?:\\.[0-9]+){0,2}\\.[0-9]+\\/tcp\\/\\d+$|^\\/ip4\\/localhost\\/tcp\\/\\d+$|^\\/ip6\\/(?:0*\\:)*?:?0*1\\/tcp\\/\\d+$/gm",
+	)
+
+	dialAddress := ""
+
+	// Find an address that's not a loopback address
+	for _, address := range addr.Addrs {
+		if !loopbackRegex.MatchString(address.String()) {
+			// Not a loopback address, dial address found
+			dialAddress = address.String()
+		}
 	}
-	return addr.Addrs[0].String() + "/p2p/" + addr.ID.String()
+
+	if dialAddress == "" {
+		panic("Unable to find appropriate dial address")
+	}
+
+	return dialAddress + "/p2p/" + addr.ID.String()
 }
 
 type PeerConnectedEvent struct {

--- a/network/server.go
+++ b/network/server.go
@@ -610,6 +610,7 @@ func AddrInfoToString(addr *peer.AddrInfo) string {
 		if !loopbackRegex.MatchString(address.String()) {
 			// Not a loopback address, dial address found
 			dialAddress = address.String()
+			break
 		}
 	}
 

--- a/network/server.go
+++ b/network/server.go
@@ -598,7 +598,7 @@ func AddrInfoToString(addr *peer.AddrInfo) string {
 	// /ip4/<any other loopback>/tcp/<port>
 	// /ip6/<any loopback>/tcp/<port>
 	loopbackRegex := regexp.MustCompile(
-		"/^\\/ip4\\/127(?:\\.[0-9]+){0,2}\\.[0-9]+\\/tcp\\/\\d+$|^\\/ip4\\/localhost\\/tcp\\/\\d+$|^\\/ip6\\/(?:0*\\:)*?:?0*1\\/tcp\\/\\d+$/gm",
+		`/^\\/ip4\\/127(?:\\.[0-9]+){0,2}\\.[0-9]+\\/tcp\\/\\d+$|^\\/ip4\\/localhost\\/tcp\\/\\d+$|^\\/ip6\\/(?:0*\\:)*?:?0*1\\/tcp\\/\\d+$/gm`,
 	)
 
 	dialAddress := ""

--- a/network/server.go
+++ b/network/server.go
@@ -597,7 +597,7 @@ var (
 	// /ip4/<any other loopback>/tcp/<port>
 	// /ip6/<any loopback>/tcp/<port>
 	loopbackRegex = regexp.MustCompile(
-		`/^\\/ip4\\/127(?:\\.[0-9]+){0,2}\\.[0-9]+\\/tcp\\/\\d+$|^\\/ip4\\/localhost\\/tcp\\/\\d+$|^\\/ip6\\/(?:0*\\:)*?:?0*1\\/tcp\\/\\d+$/gm`,
+		`^\/ip4\/127(?:\.[0-9]+){0,2}\.[0-9]+\/tcp\/\d+$|^\/ip4\/localhost\/tcp\/\d+$|^\/ip6\/(?:0*\:)*?:?0*1\/tcp\/\d+$`,
 	)
 )
 

--- a/network/server_test.go
+++ b/network/server_test.go
@@ -154,6 +154,95 @@ func TestEncodingPeerAddr(t *testing.T) {
 	assert.Equal(t, info, info2)
 }
 
+// constructMultiAddrs is a helper function for converting raw IPs to mutliaddrs
+func constructMultiAddrs(addresses []string) ([]multiaddr.Multiaddr, error) {
+	returnAddrs := make([]multiaddr.Multiaddr, 0)
+
+	for _, addr := range addresses {
+		multiAddr, err := multiaddr.NewMultiaddr(addr)
+		if err != nil {
+			return nil, err
+		}
+
+		returnAddrs = append(returnAddrs, multiAddr)
+	}
+
+	return returnAddrs, nil
+}
+
+func TestAddrInfoToString(t *testing.T) {
+	defaultPeerID := peer.ID("123")
+	defaultPort := 5000
+
+	// formatMultiaddr is a helper function for formatting an IP / port combination
+	formatMultiaddr := func(ipVersion string, address string, port int) string {
+		return fmt.Sprintf("/%s/%s/tcp/%d", ipVersion, address, port)
+	}
+
+	// formatExpectedOutput is a helper function for generating the expected output for AddrInfoToString
+	formatExpectedOutput := func(input string, peerID string) string {
+		return fmt.Sprintf("%s/p2p/%s", input, peerID)
+	}
+
+	testTable := []struct {
+		name      string
+		addresses []string
+		expected  string
+	}{
+		{
+			// A host address is explicitly specified
+			"Valid dial address found",
+			[]string{formatMultiaddr("ip4", "192.168.1.1", defaultPort)},
+			formatExpectedOutput(
+				formatMultiaddr("ip4", "192.168.1.1", defaultPort),
+				defaultPeerID.String(),
+			),
+		},
+		{
+			// Multiple addresses that the host can listen on (0.0.0.0 special case)
+			"Ignore IPv4 loopback address",
+			[]string{
+				formatMultiaddr("ip4", "127.0.0.1", defaultPort),
+				formatMultiaddr("ip4", "192.168.1.1", defaultPort),
+			},
+			formatExpectedOutput(
+				formatMultiaddr("ip4", "192.168.1.1", defaultPort),
+				defaultPeerID.String(),
+			),
+		},
+		{
+			// Multiple addresses that the host can listen on (0.0.0.0 special case)
+			"Ignore IPv6 loopback addresses",
+			[]string{
+				formatMultiaddr("ip6", "0:0:0:0:0:0:0:1", defaultPort),
+				formatMultiaddr("ip6", "::1", defaultPort),
+				formatMultiaddr("ip6", "2001:0db8:85a3:0000:0000:8a2e:0370:7334", defaultPort),
+			},
+			formatExpectedOutput(
+				formatMultiaddr("ip6", "2001:db8:85a3::8a2e:370:7334", defaultPort),
+				defaultPeerID.String(),
+			),
+		},
+	}
+
+	for _, testCase := range testTable {
+		t.Run(testCase.name, func(t *testing.T) {
+			// Construct the multiaddrs
+			multiAddrs, constructErr := constructMultiAddrs(testCase.addresses)
+			if constructErr != nil {
+				t.Fatalf("Unable to construct multiaddrs, %v", constructErr)
+			}
+
+			dialAddress := AddrInfoToString(&peer.AddrInfo{
+				ID:    defaultPeerID,
+				Addrs: multiAddrs,
+			})
+
+			assert.Equal(t, testCase.expected, dialAddress)
+		})
+	}
+}
+
 func TestJoinWhenAlreadyConnected(t *testing.T) {
 	// if we try to join an already connected node, the watcher
 	// should finish as well

